### PR TITLE
feat(rest): added filters to getAllCredentials

### PR DESCRIPTION
**Pull request by rapaktech**

### CODE REVIEW

File: packages/rest/src/controllers/credentials/CredentialController.ts [hunk 1]
```diff
@@ -31,8 +31,17 @@ export class CredentialController extends Controller {
    */
   @Example<CredentialExchangeRecordProps[]>([CredentialExchangeRecordExample])
   @Get('/')
-  public async getAllCredentials() {
-    const credentials = await this.agent.credentials.getAll()
+  public async getAllCredentials(
+    @Query('threadId') threadId?: string,
+    @Query('connectionId') connectionId?: string,
+    @Query('state') state?: string,
```
Comment (TimoGlastra) [inline packages/rest/src/controllers/credentials/CredentialController.ts:37]: Does it work if we use `CredentialState` as the type here? Maybe that'll give better validation and documentation in the swagger ui?

Comment (TimoGlastra) [reply packages/rest/src/controllers/credentials/CredentialController.ts:37]: image

File: packages/rest/src/controllers/credentials/CredentialController.ts [hunk 2]
```diff
@@ -31,8 +31,17 @@ export class CredentialController extends Controller {
...
+  ) {
+    let credentials = await this.agent.credentials.getAll()
+
+    if (state) credentials = credentials.filter((c) => c.state === state)
+    if (threadId) credentials = credentials.filter((c) => c.threadId === threadId)
```
Comment (TimoGlastra) [inline packages/rest/src/controllers/credentials/CredentialController.ts:42]: Ah I see this is also done like this in the connections controller, but we should actually filter in the query, to not retrieve records that aren't needed.

We should expose this in AFJ (I'll create an issue for this), but for now we can do it like this:

```ts
    const credentialRepository = this.agent.dependencyManager.resolve(CredentialRepository)

    const credentials = await credentialRepository.findByQuery({
      connectionId,
      threadId,
      state,
    })

    return credentials.map((c) => c.toJSON())
```

When this is merged, could you also make a PR to update this for the connections controller?

@janrtvld FYI

File: packages/rest/tests/credential.test.ts [hunk 3]
```diff
@@ -39,6 +39,48 @@ describe('CredentialController', () => {
     })
   })
 
+  describe('Get all credentials by state', () => {
+    test('should return all credentials', async () => {
```
Comment (TimoGlastra) [inline packages/rest/tests/credential.test.ts:43]: As we mock the results, we aren't testing whether te query is correctly applied here.

I think we should spy on the credential repository and see if it was called with the correct parameters:

```ts
  describe('Get all credentials by state', () => {
    test('should return all credentials', async () => {
      const credentialRepository = bobAgent.dependencyManager.resolve(CredentialRepository)
      const findByQuerySpy = jest.spyOn(credentialRepository, 'findByQuery').mockResolvedValueOnce([testCredential])

      const response = await request(app).get('/credentials').query({ state: testCredential.state })

      expect(findByQuerySpy).toBeCalledWith({
        state: testCredential.state,
      })

      expect(response.statusCode).toBe(200)
      expect(response.body).toEqual([testCredential].map(objectToJson))
    })
  })
```

**Review by TimoGlastra — CHANGES_REQUESTED** [review]
Awesome. Tested locally and it works good! Left some notes on how to make it a bit more efficient when working with large amounts of records

File: packages/rest/src/controllers/credentials/CredentialController.ts (hunk 2 above)
Comment (TimoGlastra) [reply packages/rest/src/controllers/credentials/CredentialController.ts:42]: https://github.com/hyperledger/aries-framework-javascript/issues/1022

File: packages/rest/tests/credential.test.ts (hunk 3 above)
Comment (rapaktech) [reply packages/rest/tests/credential.test.ts:43]: Got it! I'll make the changes now

**Review by TimoGlastra — APPROVED** [review]
🚀 🚀 🚀

**Review by TimoGlastra — APPROVED** [review]
🚀

**Comment by TimoGlastra** [comment]
DCO is failing for some commits (seems the last two). 

Can you sign off these commits by following these instructions: https://github.com/hyperledger/aries-rfcs/blob/main/contributing.md#how-to-sign-off-previous-commits? Then I'll merge

**Pull request merged**